src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs

```diff
@@ -633,41 +653,108 @@ private void WriteFile()
+                // Entries starting after this offset have some kind of change made to them. It might just be a fixed-length field though, in which case
+                // that single entry's metadata can be rewritten without impacting anything else.
+                long startingOffset = _firstDeletedEntryOffset;
+                long nextFileOffset = 0;
+                dynamicDirtyStartingOffset = startingOffset;
+
+                entriesToWrite = new(_entries.Count);
                 foreach (ZipArchiveEntry entry in _entries)
                 {
-                    if (!entry.LoadLocalHeaderExtraFieldAndCompressedBytesIfNeeded())
-                        markedForDelete.Add(entry);
+                    if (entry.OriginallyInArchive)
+                    {
+                        if (entry.Changed == ChangeState.Unchanged)
+                        {
+                            // Keep track of the expected position of the file entry after the final untouched file entry so that when the loop completes,
+                            // we'll know which position to start writing new entries from.
+                            nextFileOffset = Math.Max(nextFileOffset, entry.OffsetOfCompressedData + entry.CompressedLength);
+                        }
+                        // When calculating the starting offset to load the files from, only look at dirty entries which are already in the archive.
+                        else
+                        {
+                            startingOffset = Math.Min(startingOffset, entry.OffsetOfLocalHeader);
+                        }
+
+                        // We want to re-write entries which are after the starting offset of the first entry which has pending data to write.
+                        // NB: the existing ZipArchiveEntries are sorted in _entries by their position ascending.
+                        if (entry.OffsetOfLocalHeader >= startingOffset)
+                        {
+                            // If the pending data to write is fixed-length metadata in the header, there's no need to load the full file for
+                            // inflation and deflation.
+                            if ((entry.Changed & (ChangeState.DynamicLengthMetadata | ChangeState.StoredData)) != 0)
+                            {
+                                dynamicDirtyStartingOffset = Math.Min(dynamicDirtyStartingOffset, entry.OffsetOfLocalHeader);
+                            }
+                            if (entry.OffsetOfLocalHeader >= dynamicDirtyStartingOffset)
+                            {
+                                entry.LoadLocalHeaderExtraFieldAndCompressedBytesIfNeeded();
+                            }
+
+                            entriesToWrite.Add(entry);
+                        }
+                    }
+                    else
+                    {
+                        entriesToWrite.Add(entry);
```

Comment:
nit: for understandability of this long method, swap\move `entriesToWrite.Add(entry)` to the beginning of the block above (e.g. `if (!entry.OriginallyInArchive)`)